### PR TITLE
Added extension-less URLs

### PR DESCRIPTION
**Pull request by InternetExplorer7**
## Motivation

This PR also includes a patch to resolve this other change request:
#413

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Tested by re-running server to confirm page is still rendered properly and that the page is still being correctly served. I then ran 'yarn test' and that passed all test cases.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

## Change Impact Analysis
For #413, this was a server-side fix that allowed pages incorrectly classified as ".../page/index.html" to try again and render instead as ".../page.html." Now, all pages that don't include an extension still go through the same endpoint, except if there's failure, then it tries again except without adding the "/index.html" part and instead adds ".html" to the current path.
Do any existing requirements in the baseline conflict with the proposed change?
No.

Do any other pending requirements changes conflict with the proposed change?
No.

What are the business or technical consequences of not making a change?
A user using a web browser trying to access a webpage on this site that doesn't use the .html extension would be redirected to a page that doesn't render anything.

Will the proposed change adversely affect performance requirements or other quality attributes?
No.

Is the proposed change feasible within known technical constraints and current staff skills?
Yes.

Will the proposed change place unacceptable demands on any computer resources required for the development, test, or operating environments?
No, computer resources should stay the same.

Must any tools be acquired to implement and test the change?
Contributors may use the built-in test suite to test the change using command yarn test. They may also go ahead and try accessing pages with and without the .html extension.

How will the proposed change affect the sequence, dependencies, effort, or duration of any tasks currently in the project plan?
No dependencies needed to be changed and no new dependencies need to be added in order to implement this change.

Will prototyping or other use input be required to verify the proposed change?
Yes, a prototype of this webpage can be used to test if extension-less pages work and if pages with the .html extension still work as well.

How much effort that has already been invested in the project will be lost if this change is accepted?
Possibly a few hours.

Will the proposed change cause an increase in product unit cost, such as by increasing third-party product licensing fees?
No.

Will the change affect any marketing, manufacturing, training, or customer support plans?
No.

## Design Decisions

For #413, I chose to also "gut" the callback function used for the extension-less endpoint into its own function definition. I also chose to write a new function for when ".../index.html" gives an error response.

### CODE REVIEW

**Comment by JoelMarcey** [comment]
@InternetExplorer7 Hi! Sorry for the delay here. I am currently on vacation and have been over the last week. I am looking forward to reviewing this.

File: lib/server/server.js [hunk 1]
```diff
@@ -133,7 +132,7 @@ function execute(port) {
 
   app.get(/docs\/.*html$/, (req, res, next) => {
     let url = req.path.toString().replace(siteConfig.baseUrl, '');
-
+    // let url = 'docs/en/installation.html'
```
Comment (JoelMarcey) [inline lib/server/server.js]: We may have to consider the default language here. But this may be ok for now. I will have to check.

**Comment by yangshun** [comment]
@InternetExplorer7 Thanks for your contribution! Unfortunately, this isn't all you need to complete the feature.

Changing the routing in the server only works in development mode when the pages are being served via Express. In production, the site is built as static files and is served like a file directory server. After the build command is ran, th extensionless URLs wouldn't work because there's no `/path/index.html` file present when you try to access `/path`.

Try running `yarn build` and running a local webserver in the `build` directory and see for yourself. For example, for `/path/installation.html` you would need to generate the following directories and files:

```
├── path
│   ├── installation.html
│   └── installation
│       └── index.html
```

so that the web server will serve `/path/installation/index.html` when someone tries to access `/path/information`. Generating `path/installation.html` is for preserving backward compatibility.

**Comment by yangshun** [comment]
Closing due to inactivity. Feel free to reopen.

**Pull request closed without merging**